packages/notifications/src/middleware.ts

```diff
@@ -0,0 +1,70 @@
+import { Action, Dispatch, MiddlewareAPI } from 'redux';
+import {
+  TNotification,
+  TNotificationOnDismiss,
+  TNotificationAction,
+} from './types';
+import { ADD_NOTIFICATION, REMOVE_NOTIFICATION } from './action-types';
+import { removeNotification } from './action-creators';
+
+// Force TS cast of generic action to TNotificationAction
+const isNotificationAction = <Payload extends TNotification>(
+  action: Action
+): action is TNotificationAction<Payload> =>
+  [ADD_NOTIFICATION, REMOVE_NOTIFICATION].includes(action.type);
+
+const dismissCallbacksMap = new Map<
+  TNotification['id'],
+  TNotificationOnDismiss
+>();
+let id = 0;
+
+const middleware = <Payload extends TNotification>({
+  dispatch,
+}: MiddlewareAPI) => (next: Dispatch<TNotificationAction<Payload>>) => (
```

Comment:
TMiddlewareAPI or is it a external type?

Reply:
Yeah, it's from `redux`